[red-knot] Infer `tuple` types from annotations (#13943)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/generics.md

    def __init__(self, data: T):
        self.data = data

box: MyBox[int] = MyBox(5)

> # TODO should emit a diagnostic here (str and int don't unify)

Reviewer: If we're editing this message anyway, let's use language from the Python typing spec instead of Hindley-Milner language
```suggestion
# TODO should emit a diagnostic here (str is not assignable to int)
```
